### ISSUE TITLE
Allow `no-shadow` rule to not check arguments

### ISSUE THREAD

**Issue by roadhump** [eslint/eslint#2819]
It often seems better to allow `no-shadow` rule to not check argument variables.

Arguments are often need only inside function's own context, not deeper, e.g. users can use nesting  `.map(function(v, k){})` with "common" argument names `v, k` if result of parent function is not used inside of child.

`no-shadow` can give unneccessary error if user wants to make scope with `.bind`.

```
for (var i = 0; i <= 10; i++) {
    funcs.push(function(i){ // err here
    }.bind(null, i));
};
```

It slightly conflicts with `handle-callback-err` and `no-unused-vars` with `{args: 'after-used'}`
```
// eslint handle-callback-err:0
doSomething(function(err, res) {
   doAnother(res, function(err, res2) { // wish not have error here, but should add no-shadow: 0 and probably suppress real bugs

  ));
});
```

I suggest to add option to `no-shadow` rule to not check arguments. It should work something like this.

```
/*eslint no-shadow:[1, "skip-args"]*/
function some(foo, baz) {
   var bar = 1;
    return function(foo, bar) { // no err here
       var baz = 42; // err here
   };
}
```

**Comment by ilyavolodin** [comment]
@roadhump In the examples you provided, you are shadowing variables. As in inside the function, you can not access a value from the outer scope (with exception of your `bind` example). This is exactly what this rule is designed to catch. So I'm not sold on the need for this exception. I think you might want to turn this rule off, if you do not want to check function arguments.

**Comment by nzakas** [comment]
Related: #2645

**Comment by gyandeeps** [comment]
can this be closed in favor of #2645 ?

**Comment by ilyavolodin** [comment]
I don't think it's the same issue as #2645 at all. That one is talking about making an exception for function names, this one is talking about function arguments. I definitely agree with the need for #2645 I'm not sold on this one though.

**Comment by Dashed** [comment]
I'm putting in my vote for this rule enhancement. 

In some promise/generator patterns as a panacea for callback hell, `no-shadow` rule can trigger unnecessarily:

```js
co.(function* () {
    // trivial shallow pyramid for simplicity
    const {err, res} = yield new Promise(function(resolve, reject) {
            superagent
                .get(`/configs/${routeConfig}`)
                .end(function(err, res){ // `no-shadow` rule triggers here
                    resolve({err,res});
                });
        });

});

// have to rewrite as the following which isn't really idiomatic es6 code anymore :(
co.(function* () {

    const {err, res} = yield new Promise(function(resolve, reject) {
            superagent
                .get(`/configs/${routeConfig}`)
                .end(function(err2, res2){
                    resolve({err: err2, res: res2});
                });
        });

});
```

**Comment by nzakas** [comment]
@Dashed the rule is flagging those necessarily, because they are shadowing other identifiers that you no longer have access to.

But ultimately, let's just add an exception. You lose a lot of the value of the rule, but that's up to you.

**Comment by alberto** [comment]
@dashed @roadhump are you interested in submitting a PR?

**Comment by dashed** [comment]
@alberto It's been a while since I last have a need for this. Unfortunately, I'm not intimately familiar with eslint to produce the necessary and sufficient pull request.

**Comment by alberto** [comment]
Closing due to lack of activity
